File: lab/ObjcCustomTransition/ObjcCustomTransition/CustomPresentaionController.m
//
//  CustomPresentaionController.m
//  ObjcCustomTransition
//
//  Created by Lanston Peng on 7/30/14.
//  Copyright (c) 2014 Vtm. All rights reserved.
//

#import "CustomPresentaionController.h"

@interface CustomPresentaionController()

@property (strong,nonatomic)UIView* chromeView;

@end

@implementation CustomPresentaionController


- (UIView *)chromeView
{
    if(!_chromeView)
    {
        _chromeView = [[UIView alloc]initWithFrame:self.containerView.bounds];
        _chromeView.backgroundColor = [UIColor yellowColor];
        _chromeView.alpha = 0;
        _chromeView.clipsToBounds = YES;
    }
    return _chromeView;
}
- (void)presentationTransitionWillBegin
{
    for (UIView* view in self.containerView.subviews) {
        NSLog(@"%d",(int)view.tag);
    }
    [self.containerView addSubview:self.chromeView];
    [self.containerView addSubview:[self presentedView]];
    
    
    id<UIViewControllerTransitionCoordinator> transitionCoordinator = [self.presentingViewController transitionCoordinator];
    [transitionCoordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
        self.chromeView.alpha = 1;
    } completion:nil];
    
}

- (void)presentationTransitionDidEnd:(BOOL)completed
{
    if (!completed) {
        [self.chromeView removeFromSuperview];
    }
}

- (void)dismissalTransitionWillBegin
{
    id<UIViewControllerTransitionCoordinator> transitionCoordinator = [self.presentingViewController transitionCoordinator];
    [transitionCoordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
        self.chromeView.alpha = 0;
    } completion:nil];
}

- (void)dismissalTransitionDidEnd:(BOOL)completed
{
    if(completed)
    {
        [self.chromeView removeFromSuperview];
        for (UIView* view in self.containerView.subviews) {
            NSLog(@"%d",(int)view.tag);
        }
    }
}

- (CGRect)frameOfPresentedViewInContainerView
{
    return CGRectInset(self.containerView.frame, 0, 100);
}
@end

File: lab/CustomTransition/CustomTransition/CustomPresentaionController.swift
//
//  CustomPresentaionController.swift
//  CustomTransition
//
//  Created by Lanston Peng on 7/16/14.
//  Copyright (c) 2014 Vtm. All rights reserved.
//

import UIKit

class CustomPresentaionController: UIPresentationController {
    lazy var chromeView:UIView = {
        let view = UIView(frame: self.containerView!.bounds)
        view.backgroundColor = UIColor.grayColor()
        view.alpha = 0.0
        return view
    }()
    
    override func presentationTransitionWillBegin()
    {
        self.containerView.addSubview(self.chromeView)
        self.containerView.addSubview(self.presentedView())
        
        // Fade in the dimming view alongside the transition
        let transitionCoordinator = self.presentingViewController.transitionCoordinator()
        transitionCoordinator.animateAlongsideTransition({(context: UIViewControllerTransitionCoordinatorContext!) -> Void in
            self.chromeView.alpha  = 0.6
            }, completion:nil)
    }
    
    override func presentationTransitionDidEnd(completed: Bool)
    {
        if !completed
        {
            self.chromeView.removeFromSuperview()
        }
    }
    
    override func dismissalTransitionWillBegin()
    {
        let transitionCoordinator = self.presentingViewController.transitionCoordinator()
        transitionCoordinator.animateAlongsideTransition({(context:UIViewControllerTransitionCoordinatorContext!) -> Void in
            self.chromeView.alpha = 0.2;
            }, completion: nil)
    }
    override func dismissalTransitionDidEnd(completed: Bool)
    {
        if completed {
            self.chromeView.removeFromSuperview()
        }
    }
    
    
    
    override func frameOfPresentedViewInContainerView() -> CGRect {
        // We don't want the presented view to fill the whole container view, so inset it's frame
        var frame = self.containerView.bounds;
        frame = CGRectInset(frame, 0 , 100.0)
        return frame
    }

}
